tasks/EchoesToFrameConverterTask.cpp

```diff
@@ -0,0 +1,172 @@
+    while (_echo.read(radar_echo, false) == RTT::NewData) {
+        if (m_current_sweep_size != radar_echo.sweep_length ||
+            m_current_num_angles != (2 * M_PI / radar_echo.step_angle.getRad()) ||
+            m_current_range != radar_echo.range) {
+            m_current_sweep_size = radar_echo.sweep_length;
+            m_current_num_angles = 2 * M_PI / radar_echo.step_angle.getRad();
+            m_current_range = radar_echo.range;
+            updateLookUpTable();
+            m_echoes.clear();
+        }
+
+        addEchoesToFrame(radar_echo, yaw_correction);
+    }
+    if (base::Time::now() - m_last_sample > _export_config.get().time_between_frames) {
+        publishFrame();
+    }
+}
+void EchoesToFrameConverterTask::errorHook()
+{
+    EchoesToFrameConverterTaskBase::errorHook();
+}
+void EchoesToFrameConverterTask::stopHook()
+{
+    EchoesToFrameConverterTaskBase::stopHook();
+}
+void EchoesToFrameConverterTask::cleanupHook()
+{
+    EchoesToFrameConverterTaskBase::cleanupHook();
+}
+
+void EchoesToFrameConverterTask::updateLookUpTable()
+{
+    LOG_INFO_S << "Updating LookUpTable";
+    auto config = _export_config.get();
+    m_lut.reset(new EchoToImageLUT(m_current_num_angles,
+        m_current_sweep_size,
+        config.beam_width,
+        config.window_size));
+}
+
+void EchoesToFrameConverterTask::addEchoesToFrame(Radar const& echo,
+    double yaw_correction)
+{
+    LOG_INFO_S << "Adding Echoes to frame";
+    double angle = echo.start_heading.getRad() + yaw_correction;
+    if (angle < 0) {
+        angle = 2 * M_PI + angle;
+    }
+
+    int start_angle_unit = round(angle / echo.step_angle.getRad());
+    int angles_in_a_frame = 2 * M_PI / echo.step_angle.getRad();
+    LOG_INFO_S << "ANGLE MESSAGE-> start: " << start_angle_unit
+               << "YAW Correction: " << yaw_correction
+               << " end: " << start_angle_unit + echo.sweep_timestamps.size()
+               << " Sweep length: " << echo.sweep_length;
+    for (int current_angle = 0;
+         current_angle < static_cast<int>(echo.sweep_timestamps.size());
+         current_angle++) {
+        LOG_INFO_S << "ANGLE LOOP-> Index: " << current_angle << " angle message: "
+                   << start_angle_unit + current_angle % angles_in_a_frame << "/"
+                   << angles_in_a_frame << " sweep: " << current_angle * echo.sweep_length
+                   << "/" << (current_angle + 1) * echo.sweep_length
+                   << " total sweep length: " << echo.sweep_data.size();
+
+        m_echoes[(start_angle_unit + current_angle) % angles_in_a_frame] =
+            std::make_unique<std::vector<uint8_t>>(echo.sweep_data.begin() +
+                                                       current_angle * echo.sweep_length,
+                echo.sweep_data.begin() + (current_angle + 1) * echo.sweep_length);
+    }
+}
+
+int EchoesToFrameConverterTask::discretizeAngle(double theta_rad, int num_angles)
+{
+    while (theta_rad < 0) {
```

Comment:
You test < 0 but not  > 2 * PI. This is inconsistent (and therefore weird).

Anyways. Use base::Angle for your angle needs (including passing it to this method), which guarantee that the angle is in ]-PI, PI]. Then you can just do a `if (rad < 0) { rad += M_PI; }`

Reply:
I'm no longer using this function so I removed it.